File: uva online judge/00341 - Non-Stop Travel.py
import math

def read_test_case():
    n = int(input())
    if n == 0:
        return None
    adj_matrix = [[math.inf] * n for i in range(n)]
    for i in range(0, n):
        nums = [int(j) for j in input().split()]
        for j in range(nums[0]):
            adj_matrix[i][nums[2 * j + 1] - 1] = min(nums[2 * j + 2], adj_matrix[i][nums[2 * j + 1] - 1])
    start, end = (int(i) - 1 for i in input().split())
    input()
    return n, adj_matrix, start, end

def floyd_warshall(n, adj_matrix):
    parent = [[i] * n for i in range(n)]
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if adj_matrix[i][k] + adj_matrix[k][j] < adj_matrix[i][j]:
                    adj_matrix[i][j] = adj_matrix[i][k] + adj_matrix[k][j]
                    parent[i][j] = parent[k][j]
    return  adj_matrix, parent

def print_path(parent, start, end):
    if start != end:
        print_path(parent, start, parent[start][end])
    print(' {}'.format(end + 1), end='')
    
if __name__ == '__main__':
    for i, (n, adj_matrix, start, end) in enumerate(iter(read_test_case, None), start=1):
        adj_matrix, parent = floyd_warshall(n, adj_matrix)        
        print('Case {}: Path ='.format(i), end='') 
        print_path(parent, start, end)
        print('; {} second delay'.format(adj_matrix[start][end]))

File: uvaonlinejudge/00341 - Non-Stop Travel.c
#include <stdio.h>
#include <limits.h>
#define MAX_NODES 10
#define INF INT_MAX

int adj_matrix[MAX_NODES][MAX_NODES], parent[MAX_NODES][MAX_NODES];

int read_test_case(int *n, int adj_matrix[][MAX_NODES], int *start, int *end){
    int i, j, v, weight;
    scanf("%d", n);
    if(*n == 0)
        return 0;
    for(i = 0; i < *n; i++)
        for(j = 0; j < *n; j++)
            adj_matrix[i][j] = INF;       
    for(i = 0; i < *n; i++){
        scanf("%d", &j);
        for(; j > 0; j--){
            scanf("%d %d", &v, &weight);
            if(weight < adj_matrix[i][v - 1]){
                adj_matrix[i][v - 1] = weight;
            }
        }
    }
    scanf("%d %d", start, end);
    (*start)--; (*end)--;
    return 1;
}

void floyd_warshall(int n, int adj_matrix[][MAX_NODES], int parent[][MAX_NODES]){
    int i, j, k;
    for(i = 0; i < n; i++)
        for(j = 0; j < n; j++)
            parent[i][j] = i;
    for(k = 0; k < n; k++){
        for(i = 0; i < n; i++){
            for(j = 0; j < n; j++){
                if(adj_matrix[i][k] != INF &&  adj_matrix[k][j] != INF && adj_matrix[i][j] > adj_matrix[i][k] + adj_matrix[k][j]){
                    adj_matrix[i][j] = adj_matrix[i][k] + adj_matrix[k][j];
                    parent[i][j] = parent[k][j];
                }
            }
        }
    }
}

void print_path(int parent[][MAX_NODES], int start, int end){
    if( start != end)
        print_path(parent, start, parent[start][end]);
    printf(" %d", end + 1);
}

int main(){
    int i = 1, n, start, end ;
    while(read_test_case(&n, adj_matrix, &start, &end)){
        floyd_warshall(n, adj_matrix, parent);
        printf("Case %d: Path =", i);
        print_path(parent, start, end);
        printf("; %d second delay\n", adj_matrix[start][end]);
        i++;
    }
    return 0;
}